### ISSUE TITLE
Should we temporarily disable quarantining

### ISSUE THREAD

**Issue by vitorgalvao** [Homebrew/homebrew-cask#70798]
Quarantining seems to be the cause of a ton of issues these days, be it problems with `xattr` or with Catalina. From what I can gather so far, Catalina and notarisation requirements are little more than security theatre, as removing (or in our case, not adding) the quarantine attribute bypasses the whole security checks system. [A blog post by Jeff Johnson](https://lapcatsoftware.com/articles/notarization.html) seems to confirm this (scroll down to the last section: “False Benefit: Malware Scan”).

Manually adding the quarantine attribute is a somewhat opaque process (@reitermarkus did research on this early on) and Apple doesn’t provide a tool to make it simple, so we had to devise our own (implemented in Swift by @amyspark), bending over backwards to accommodate Apple’s Gatekeeper security model.

However, it has become abundantly clear that something changed on Apple’s side, and our solution is creating issues for our users — too many for us to keep up with until we get everything straight.

Thus, as a proponent *for* macOS security, I sadly propose we disable the quarantining feature for now, until we can either recover our footing or Catalina’s system changes. We may wish to include a message on every `install`, informing users quarantining is no longer in place, lest they get a false sense of security.

Opening this post to gather opinions from maintainers and users alike.

Pinging @Homebrew/cask.

**Comment by amyspark** [comment]
I do agree that Catalina has made macOS security a mess. The real problem is that **Apple loves changing tools without detailing them in changelogs**. The cornerstone of Gatekeeper is extended attributes, and the only tool to handle them in macOS is `xattr`.

`xattr` has two main issues:

- Apple loves taking away functionality. When Mojave was released, Apple took away its `-R` (recursive) flag, and with Catalina, it started following symlinks blindly. And in neither case there was a changelog.
- It depends on system Python, which is *very* vulnerable with SIP disabled. All issues of "distribution not found", "xattr==0.6.4", "pkg_resources", etc. come up after someone did so and tried installing Python from a pkg.

in re security -- I do not have any problem with Apple introducing whatever changes they deem necessary, but I just want some headsup. 

As for the rest, FTR, I think it's high time we started moving away from this tool in particular. If we keep getting this kind of bug reports, I want to be able to pull the trigger and remove `xattr` usage from Homebrew altogether. I have got a proof of concept in Swift to deal with it.

**Comment by vitorgalvao** [comment]
> It depends on system Python

Does that mean there’s a good chance `xattr` will be killed off in a future release on macOS, when they no longer provide binaries for languages such as Python and Ruby?

> which is _very_ vulnerable with SIP disabled.

If a user’s system was modified as part of disabling SIP, I don’t think we should offer support for it. Just like (if I recall correctly) HB doesn’t support installations to non-standard locations, disabling SIP is *strongly* advised against, is done by a small number of users, and can lead to unforeseen consequences. If someone disables SIP, they should be doing so at their own peril and be ready to deal with the consequences.

I think that in some of those issues, the `xattr issues` section reported `<NONE>`. How are we checking for those? Could it be useful to (for example) take hashes of both `xattr` and the system Python and compare them to see if any were modified?

**Comment by amyspark** [comment]
`xattr issues` displays an error if and only the `xattr` executable is broken, no if our expectations no longer match what it does. (Like it just happened with symlinks.)

**Comment by vitorgalvao** [comment]
> if our expectations no longer match what it does.

Wouldn’t that be a better check? Have `xattr issues` do a little test to see if it’s behaving as we expect?

**Comment by core-code** [comment]
as an app vendor, i don't want my apps quarantined, especially on Catalina.

there seems to be a bug in 'Catalina' where it will randomly crash our app if it is launched under translocation/quarantine. we've received dozens of crash reports there. Apple does not want to fix this (or seemingly, any other) issue. 

the only way to prevent these crashes is to make sure that translocation never occurs. this means changing from ZIP delivery to signed disk images. furthermore, our app will be changed to refuse to launch if it is is launched with quarantine, as it might just crash randomly later on. 

will HBC put apps under quarantine even if they are delivered on signed disk images?

**Comment by vitorgalvao** [comment]
> will HBC put apps under quarantine even if they are delivered on signed disk images?

I think we don’t differentiate and just quarantine whatever we download.

But I’ve noticed (and this seems to be a feature, not a bug) that when installing a `pkg` via the command-line with `installer` (which we do), Gatekeeper checks are skipped. That means our method of deferring security checks to Apple doesn’t work on one of the cases where it may matter most.

This is frustrating because I no longer feel like we can count on Apple to do the right thing. Their privacy and security protections are lacklustre and buggy, and the honest developers and users are the ones being punished by their lack of care (or competence?).

I’m open to changing our system. For example, instead of quarantining downloads we could run `spctl --assess` on relevant artifacts (`app`, `pkg`, …) and skip the installation (and inform the user why) if they don’t exit with `0`. We’d also replace `--no-quarantine` with `--no-spctl-check` or similar.

**Comment by core-code** [comment]
> spctl --assess on relevant artifacts (app, pkg, …) 

which won't help if e.g. there is something malicious in there but its properly signed.

why not store the expected code signature author in a new stanza?

this would help
1.) when submitting an update (e.g. because of a SHA mismatch) travis would should immediately if the signature is not as expected and
2.) for casks that have :no_check the user would still be protected because 'cask install' would only do it if the signature is the right one.

obviously that wont help if an attacker has access to the private keys of the vendor, but i don't think anything will help in this case.

**Comment by vitorgalvao** [comment]
> which won't help if e.g. there is something malicious in there but its properly signed.

True, but neither will the current system.

> why not store the expected code signature author in a new stanza?

That would require a non-trivial amount of extra work<sup>1</sup>. I’m not opposed to it, but I’d prefer to have something simpler in place for at least one full macOS release, to make sure we’re happy with it before adding more.

<sup>1</sup> I’m getting flashbacks of the `license` stanza. Those were not happy times.

**Comment by core-code** [comment]
> That would require a non-trivial amount of extra work

thats always a good argument ;) 

though, it is possible to automatically extract the code signature from a cask, while this is not possible with the license. it could be pretty well automated.

**Comment by vitorgalvao** [comment]
> it is possible to automatically extract the code signature from a cask, while this is not possible with the license. it could be pretty well automated.

Fair, I agree.

But my immediate concern is with having to check every cask one by one and add the stanza. Large-scale changes like that tend to be problematic and bring idiosyncrasies and edge-cases to the surface all at once, which is why I suggest the waiting period between steps. If we implement the simpler solution but find it inadequate, we save the effort of going into the next phase and checking, adding, and removing the stanza from every cask.

**Comment by core-code** [comment]
adding the stanza could be automated as well i guess. 

but i agree that we can't do any big changes in the near term and should investigate simpler options first.

**Comment by mtibben** [comment]
I'm not convinced disabling quarantining will solve much. Anyone wanting to offer downloads to run on Catalina will need to jump through the signing and notorization hoops anyway. It's not just cask who will face these issues, upstream will too.

**Comment by core-code** [comment]
> I'm not convinced disabling quarantining will solve much. 

however, if Apple itself makes sure that an app installed from a signed disk image into the /Applications folder does not have the quarantine flag set, then so should HBC

> Anyone wanting to offer downloads to run on Catalina will 
> need to jump through the signing and notarisation hoops anyway.

yes and in reward he can expect his app not to be quarantined once it arrives at the user. 
unless that user uses HBC, in which case it is STILL quarantined. this is not in the interest of the vendor, or the user if the app does not work as expected due to the quarantine.

**Comment by vitorgalvao** [comment]
> Anyone wanting to offer downloads to run on Catalina will need to jump through the signing and notorization hoops anyway.

That’s incorrect. There are tons of macOS apps that are and will continue to be unsigned as long as Apple charges $99/year. From the perspective of a developer who didn’t sign their apps before, nothing as changed—their users still need to right-click and `Open` to run the app.

> It's not just cask who will face these issues, upstream will too.

Thousands of people use only HBC to install their apps. Giving them a bad experience because they would have hypothetically gotten one if they had downloaded from the vendor’s website isn’t a strong argument.

**Comment by mtibben** [comment]
> Apple itself makes sure that an app installed from a signed disk image into the /Applications

Ahhh I didn't realise that. Is that documented anywhere? Is there any other location or circumstances where the quarantine flag is removed? I've been [running into this issue myself](https://github.com/99designs/aws-vault/issues/430#issuecomment-546883740) packaging aws-vault

@vitorgalvao:
> Thousands of people use only HBC to install their apps. Giving them a bad experience because they would have hypothetically gotten one if they had downloaded from the vendor’s website isn’t a strong argument.

Removing the quarantine flag bypasses Gatekeeper's checks altogether. Does HBC want to put itself in the situation of allowing potentially malicious code to run that macOS would have prevented? My argument is that if upstream is forced to deal with this issue anyway, why would HBC short-circuit this?

**Comment by vitorgalvao** [comment]
> Does HBC want to put itself in the situation of allowing potentially malicious code to run that macOS would have prevented?

* That’s how HBC worked for the majority of its existence.
* We’ve always been straightforward that the user is the last line of defense, and have been explicit that you shouldn’t fall into a false sense of security and think “because it’s on HBC, it’s safe”. We do what we can, but malicious-software-hunting is not a core goal. Though we welcome @core-code’s [malware checks](https://github.com/Homebrew/homebrew-cask/issues/56223).
* Quarantining doesn’t prevent against malicious pkgs (because we need to use `install`, which ignores it), which have the most destructive capabilities: an app you still have to run to be infected, but a pkg does it at install time.

There is [a suggestion](https://github.com/Homebrew/homebrew-cask/issues/70798#issuecomment-546368051) about what to replace quarantining with, with [possible next steps](https://github.com/Homebrew/homebrew-cask/issues/70798#issuecomment-546369765). It:

* Removes a source of frequent bug reports.
* Leaves users *more* informed about what they’re installing.
* Makes it less annoying and immediate when they do want to install (issue is resolved at install time, not first launch).
* Removes one dependency on an external programming language.

> My argument is that if upstream is forced to deal with this issue anyway, why would HBC short-circuit this?

There’s a significant part of upstream who *won’t* deal with this, and *we* will be the ones stuck with the bug reports.

This is not a simple decision of turning quarantining on or off; having it on has costs for us. Apple could ease that cost with ease if they provided a tool to quarantine things with ease. Yes, I’ve opened a radar; no, they never replied.

**Comment by mtibben** [comment]
Cool, thanks for the detailed answers. Much clearer now

**Comment by pvande** [comment]
For an “outsider’s” perspective...

I stumbled across this issue while investigating the feasibility of automatically bypassing Gatekeeper.  I’m using Homebrew Bundle to specify a set of baseline applications when I bootstrap a new machine.  **For this workflow, working to support Gatekeeper is unhelpful.**. While the possibility exists that HBC will install a compromised version of, say, Chrome, no roadblock from Gatekeeper is going to help me notice that – I asked for Chrome, I got Chrome, and all Gatekeeper will tell me is that Chrome came from HBC.

On the other hand, there’s the workflow where I’m manually doing exploratory installs ad hoc.  Gatekeeper is *potentially* more useful here, in that it provides an opportunity to visually validate that HBC downloaded the right [icon], but the lack of package support and the general weakness of the contract (again, I asked for X, got X, GK tells me X came from HBC) isn’t adding much real security.

Then there’s the case where HBC was invoked by another [untrusted] application or script, and not directly (or perhaps knowingly) by the user.  **This is the best case I can come up with for GK support, as it signifies something unusual.**  Even so, with `install` being compromised, a well-written malicious script wouldn’t run afoul of GK anyway...

---

My own take is that GK support isn’t adding much, except a false sense of security and maintenance overhead.  It may well not be worth supporting.

If HBC *does* continue to support it, I would be in favor of adding a flag to specify an `spctl --label` for the application.  That would allow users to configure their own “trusted” namespace (as described [here](http://www.theinstructional.com/guides/gatekeeper-fundamentals-part-2)), but leave the default behavior in place.

**Comment by vitorgalvao** [comment]
> Then there’s the case where HBC was invoked by another [untrusted] application or script, and not directly (or perhaps knowingly) by the user. **This is the best case I can come up with for GK support, as it signifies something unusual.**

That’s not a realistic attack. If an external tool can invoke HBC, it can also get the software directly through `curl` and run it the same way we do. We use available system tools and don’t have any special permissions. In addition, [the above suggestion](https://github.com/Homebrew/homebrew-cask/issues/70798#issuecomment-546368051) would be more effective at preventing such attacks.

> Even so, with `install` being compromised

My mistake above: it’s `installer`. But I not sure “compromised” is the right word; this looks to be a feature.

> If HBC _does_ continue to support it, I would be in favor of adding a flag to specify an `spctl --label` for the application.

Use `--no-quarantine`. It’s already in place and it’s a simpler way to achieve the same thing.

**Comment by pvande** [comment]
> That’s not a realistic attack. If an external tool can invoke HBC, it can also get the software directly through curl and run it the same way we do. We use available system tools and don’t have any special permissions. In addition, the above suggestion would be more effective at preventing such attacks.

Agreed on all points, especially given the `--no-quarantine` option.

> But I not sure “compromised” is the right word[…]

To be clear, I meant "compromised" in the sense that it is a discontinuity in the Gatekeeper security front.  In that sense, I believe it can be both a deliberate feature of `installer`, and a compromise of HBC.

> Use --no-quarantine. 

Great!  I hadn't found that in my searching.

---

Overall, this leaves me even more convinced that supporting GK isn't particularly worthwhile.

**Comment by brandonkal** [comment]
To add to the discussion, I am also experiencing issues here #72421

**Comment by vitorgalvao** [comment]
Another example of where the current method causes issues: https://github.com/Homebrew/homebrew-cask/issues/72813.

The way to fix that with the current method is to change the installation method to `manual`, which makes installation a worse experience.

**Comment by vitorgalvao** [comment]
Another example: https://github.com/Homebrew/homebrew-cask/issues/70798.

**Comment by core-code** [comment]
> Another example: #70798.

wrong pr number? ;)


in any case, it really looks like the quarantining should be disabled.

how about showing some info about the code-signature after installation instead:
1.) print info if there is code signing at all
1.1.) if yes, print the name of the issuer
1.2.) if yes, also print whether the app is additionally notarized or sandboxed
1.3.) if the signature does not check out (`spctl --verbose=4 --raw --assess --type execute APP &&  codesign --verify --verbose=4 --deep --strict APP`), print some red error message and/or fail installation

**Comment by vitorgalvao** [comment]
> wrong pr number?

Meant https://github.com/Homebrew/homebrew-cask/pull/73145. Fixed. Thank you.

**Comment by vitorgalvao** [comment]
Another issue: https://github.com/Homebrew/homebrew-cask/issues/73752

**Comment by zcutlip** [comment]
I don't think there's a cask stanza for an individual cask to opt out of quarantine. Maybe there should be?

**Comment by vitorgalvao** [comment]
> I don't think there's a cask stanza for an individual cask to opt out of quarantine. Maybe there should be?

There shouldn’t, no. We don’t get quarantining for free, we [had to do a ton of work](https://github.com/Homebrew/homebrew-cask/issues/70798#issue-506738809) to add it. Implementing a security feature and unilaterally selectively disabling it when it’s inconvenient makes little sense. It would not fix the deeper issue, would require more maintenance, would be confusing, and would require users to always be checking over their shoulder.

What we should do is either fix the system we have (which is not trivial and would still [leave a gaping hole regarding `pkg`s](https://github.com/Homebrew/homebrew-cask/issues/70798#issuecomment-546368051)), [replace it](https://github.com/Homebrew/homebrew-cask/issues/70798#issuecomment-546368051), or get rid of it.

**Comment by zcutlip** [comment]
That's a fair point.

What about a new caveat method that gives a warning about the quarantine attribute and auto-generates the appropriate `xattr` command the user can run to remove the quarantine attribute if/when they're comfortable with the integrity of the installed cask?

So if one applied:
```
caveats do
  quarantine
end
```
Output might look like:
```
==> Caveats
The quarantine attribute has been applied to the jtool2 cask.
Because the casks's binaries are not properly code signed, they will not execute until the the quarantine attribute has been removed.
You can remove the quarantine attribute from this cask with the following command:
xattr -r -d com.apple.quarantine /usr/local/Caskroom/jtool2
```

**Comment by vitorgalvao** [comment]
> What about a new caveat method that gives a warning about the quarantine attribute and auto-generates the appropriate `xattr` command the user can run to remove the quarantine attribute if/when they're comfortable with the integrity of the installed cask?

That doesn’t fix the problem, it just adds another workaround that will make everyone unhappy:

* Frustrated maintainers and collaborator due to the extra work of needing to keep track of signing status of every piece of software.
* Frustrated users who will complain that they have to manually run a command all the time, negating the point of HBC.

We don’t want workarounds; [workarounds are bad](https://github.com/Homebrew/homebrew-cask/issues/13201). We need to [fix the system, change it, or remove it](https://github.com/Homebrew/homebrew-cask/issues/70798#issuecomment-571870925).

**Comment by vitorgalvao** [comment]
@Homebrew/cask I’d like further feedback, here. I think we should go with [the simpler replacement strategy](https://github.com/Homebrew/homebrew-cask/issues/70798#issuecomment-546368051) and consider [something more complex](https://github.com/Homebrew/homebrew-cask/issues/70798#issuecomment-546369765) if the former proves to be effective.

If no one wants to or can tackle that now, let’s at least get rid of the current system first. I’ll appreciate help / a PR there, but if no one can do it I’ll take a stab. It’s a shame to see our effort go to waste there, but it should not be our burden to work around Apple’s lack of commitment to software quality and supporting developers who want to keep users secure. It would be a different case if we felt we could count on them to get things fixed, but it’s becoming abundantly clear they don’t give a hoot.

**Comment by jgogstad** [comment]
What about:
1. Default behavior: Set the quarantine flag if and only if the cask is unsigned or signature does not validate
1. Keep existing opt-out with `--no-quarantine`
1. Optionally add a flag to restore current behavior

The above list can also be interpreted as add an option `--quarantine=(always|never|unsigned)` and have it default to `unsigned`.

**Comment by vitorgalvao** [comment]
@jgogstad That doesn’t fix the issues (I’m on a phone and would rather not keep repeating them), is confusing, and requires extra work.

**Comment by reitermarkus** [comment]
@vitorgalvao, so the simple replacement would be the `spctl --assess` check?

I definitely agree we should remove quarantining until there's a proper way we can support it, which I'm not holding my breath for.

**Comment by vitorgalvao** [comment]
> @vitorgalvao, so the simple replacement would be the `spctl --assess` check?

Yes, [the last paragraph](https://github.com/Homebrew/homebrew-cask/issues/70798#issuecomment-546368051):

> (…) run `spctl --assess` on relevant artifacts (`app`, `pkg`, …) and skip the installation (and inform the user why) if they don’t exit with `0`. We’d also replace `--no-quarantine` with `--no-spctl-check` or similar.

Though we could open an issue to discuss the interaction with the user.

We could do something else, but that seems like the simplest solution while being effective, and starting simple is important—no point in doing something complex now if the base doesn’t work.

> I definitely agree we should remove quarantining until there's a proper way we can support it, which I'm not holding my breath for.

Same. A proper way would require some kind of Apple support, and we can’t even get proper documentation.

**Comment by vitorgalvao** [comment]
[Another one](https://github.com/Homebrew/homebrew-cask/issues/78946).

**Comment by schwedenmut** [comment]
Happens also on https://github.com/Homebrew/homebrew-cask/blob/master/Casks/burp-suite.rb

**Comment by beeuhtricks** [comment]
Is there any way I could work around this temporarily? It looks like cask is deleting burp-suite before I can bypass the quarantine in System Preferences.

**Comment by vitorgalvao** [comment]
> Is there any way I could work around this temporarily?

Use `--no-quarantine` as a flag when installing.

**Comment by hamiltont** [comment]
Just mentioning this in case it helps discussion: 

There is a gatekeeper exclusion list at `Prefs>Security&Privacy>Privacy>Developer Tools`. AFAIK, if you add your terminal of choice to that list, anything launched from terminal also inherits the exclusion. That may not be what everyone wants, given we've just flipped a big security toggle to OFF, but maybe it would help some folks here

**Comment by vitorgalvao** [comment]
@hamiltont We go to great lengths to turning quarantining *on*. We don’t need hacks to turn it off (it’s not added by `curl`, which Homebrew uses), we needed hacks to turn it on.

**Comment by claui** [comment]
Don’t mind me, just leaving a link here for later reference when I find time to tackle the issue at hand:

https://eclecticlight.co/2020/07/10/how-macos-overrides-app-behaviour-including-quarantine/

It doesn’t offer a solution but explains a few pieces of the underlying behaviour.

**Comment by vitorgalvao** [comment]
[One more](https://github.com/Homebrew/homebrew-cask/issues/86111).

**Comment by miccal** [comment]
[And another one.](https://github.com/Homebrew/homebrew-cask/issues/87415)

**Comment by vitorgalvao** [comment]
> [And another one.](https://github.com/Homebrew/homebrew-cask/issues/87415)

That one is working as intended. Quarantining was applied correctly, it’s just that the app isn’t signed. This issue is for cases where our method of adding quarantining causes problems.

**Comment by miccal** [comment]
Ah, ok.

**Comment by miccal** [comment]
Sorry @vitorgalvao :(

**Comment by vitorgalvao** [comment]
> Sorry @vitorgalvao :(

For an unimportant oversight amidst an onslaught of stellar contributions and help to other users? There’s absolutely nothing to apologise for.

**Comment by miccal** [comment]
> For an unimportant oversight amidst an onslaught of stellar contributions and help to other users? There’s absolutely nothing to apologise for.

Thanks for the kind words :)

**Comment by 0xced** [comment]
For reference: I have added `export HOMEBREW_CASK_OPTS="--no-quarantine"` in my `~/.zshrc` file so that the `--no-quarantine` is automatically added everytime I install a cask. This gets rid of all quarantine-related issues and gets rid of these annoying dialogs popups: *This is an application downloaded from the Internet. Are you sure you want to open it?*

**Comment by dcondrey** [comment]
I noticed this issue when I tried to install `teamviewer` this evening.

Attempting to run `brew install teamviewer` would fail with a Error: `...TeamViewer.dmg was not quarantined properly.` because I have this in my `.zshrc`:

    HOMEBREW_CASK_OPTS="--appdir=/Applications --fontdir=/Library/Fonts --no-quarantine"

But when I explicitly run `brew install teamviewer --no-quarantine` I was prompted for my admin password and it installed successfully.  For some reason having it set in the variable I was not being prompted for admin password.
